File: 0079-word-search/0079-word-search.py
class Solution(object):
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        rows = len(board)
        cols = len(board[0])
        
        def backtrack(ind, i, j):
            if ind == len(word):
                return True
            if i < 0 or i >= rows or j < 0 or j >= cols or board[i][j] != word[ind]:
                return False
            
            temp = board[i][j]
            board[i][j] = ""
            
            found = (backtrack(ind+1, i+1, j) or backtrack(ind+1, i-1, j) or backtrack(ind+1, i, j+1) or backtrack(ind+1, i, j-1))
            
            board[i][j] = temp
            return found
            
            
        for i in range(rows):
            for j in range(cols):
                if board[i][j] == word[0]:
                    if backtrack(0, i, j):
                        return True
                    
        return False

File: 0079-word-search/0079-word-search.java
class Solution {
    public boolean exist(char[][] board, String word) {
        int rows = board.length;
        int cols = board[0].length;
        int len = word.length();
        
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (backtrack(board, word, i, j, 0)) {
                    return true;
                }
            }
        }
        
        return false;
    }
    
    private boolean backtrack(char[][] board, String word, int i, int j, int k) {
        int rows = board.length;
        int cols = board[0].length;
        
        if (k == word.length()) {
            return true;
        }
        
        if (i < 0 || i >= rows || j < 0 || j >= cols || board[i][j] != word.charAt(k)) {
            return false;
        }
        
        char temp = board[i][j];
        board[i][j] = '.';
        
        boolean found = backtrack(board, word, i + 1, j, k + 1) || 
                        backtrack(board, word, i - 1, j, k + 1) || 
                        backtrack(board, word, i, j + 1, k + 1) || 
                        backtrack(board, word, i, j - 1, k + 1);
        
        board[i][j] = temp;
        
        return found;
    }
}